pow/src/main/java/tech/pegasys/artemis/pow/DepositProcessingController.java

```diff
@@ -0,0 +1,210 @@
+  }
+
+  private synchronized void onSubscriptionFailed(Throwable err) {
+    Disposable subscription = newBlockSubscription;
+    if (subscription != null) {
+      subscription.dispose();
+    }
+    LOG.warn("New block subscription failed, retrying.", err);
+    asyncRunner
+        .getDelayedFuture(Constants.ETH1_SUBSCRIPTION_RETRY_TIMEOUT, TimeUnit.SECONDS)
+        .finish(
+            () -> startSubscription(latestSuccessfullyQueriedBlock),
+            (error) ->
+                LOG.warn(
+                    "Unable to subscribe to the Eth1Node. Node won't "
+                        + "have access to new deposits after genesis.",
+                    error));
+  }
+
+  private synchronized void onNewCanonicalBlockNumber(BigInteger latestCanonicalBlockNumber) {
+    if (latestCanonicalBlockNumber.compareTo(this.latestCanonicalBlockNumber) <= 0) {
+      return;
+    }
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    fetchLatestSubscriptionDeposits();
+  }
+
+  private synchronized void fetchLatestSubscriptionDeposits() {
+    if (isBlockByBlockModeOn) {
+      fetchLatestDepositsOneBlockAtATime();
+    } else fetchLatestSubscriptionDepositsOverRange();
+  }
+
+  private synchronized void fetchLatestSubscriptionDepositsOverRange() {
+    final BigInteger toBlock;
+    final BigInteger fromBlock;
+
+    if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+      return;
+    }
+    active = true;
+
+    fromBlock = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    toBlock = latestCanonicalBlockNumber;
+
+    depositFetcher
+        .fetchDepositsInRange(fromBlock, toBlock)
+        .finish(
+            __ -> onSubscriptionDepositRequestSuccessful(toBlock),
+            (err) -> onSubscriptionDepositRequestFailed(err, fromBlock, toBlock));
+  }
+
+  private synchronized void fetchLatestDepositsOneBlockAtATime() {
+    final BigInteger nextBlockNumber;
+
+    synchronized (DepositProcessingController.this) {
+      if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+        return;
+      }
+      active = true;
+
+      nextBlockNumber = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    }
+
+    depositFetcher
+        .fetchDepositsInRange(nextBlockNumber, nextBlockNumber)
+        .thenCompose(
+            __ ->
+                eth1Provider.getGuaranteedEth1BlockFuture(
+                    UnsignedLong.valueOf(nextBlockNumber), asyncRunner))
+        .thenAccept(
+            block -> {
+              final BigInteger blockNumber = block.getNumber();
+              LOG.trace("Successfully fetched block {} for min genesis checking", blockNumber);
+              LOG.trace(
+                  "Seconds until min genesis block {}",
+                  Constants.MIN_GENESIS_TIME.minus(
+                      calculateCandidateGenesisTimestamp(block.getTimestamp())));
```

Comment:
This should probably be in an else block or the minus will underflow when min genesis time has been reached.

Reply:
Makes sense.